### PR TITLE
Scheduler

### LINKED ISSUE #10: write a scheduler using java threading
- [x] Start an async thread execution on an api call event
- [x] New thread should run a task periodically 
- [x] stop the thread after some set of execution

### PR DESCRIPTION
**Pull request by Poonam00**
fixes #10

### CODE REVIEW

File: myworkspacestaff_managementlog.txt [hunk 1]
```diff
@@ -0,0 +1,355 @@
+2020-05-17 20:30:55.654  INFO 8220 --- [restartedMain] c.s.StaffManagementApplication           : Starting StaffManagementApplication on DESKTOP-SLLSFPL with PID 8220 (E:\myworkspace\staff_management\staff-management-api\target\classes started by poonam.sharma in E:\myworkspace\staff_management\staff-management-api)
```
Comment (rvashishth) [inline myworkspacestaff_managementlog.txt]: remove this file. i would suggest to add this in `.gitignore` after deleting from repo

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 2]
```diff
@@ -1,16 +1,52 @@
...
 import org.springframework.web.bind.annotation.RestController;
 
+import com.staffmanagement.service.ScheduleService;
+
 @RestController
+@ControllerAdvice
 public class UtilityController {
 
-	@GetMapping("/getHeader")
+	Logger log = LoggerFactory.getLogger(UtilityController.class);
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]: use accessibility identifiers. i.e. Logger should be `private`

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 3]
```diff
@@ -1,16 +1,52 @@
...
+
+	@Autowired
+	ScheduleService service;
+
+	@GetMapping("/header")
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]: as this api return map of headers, we better name this `/headers` in the plural

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 4]
```diff
@@ -1,16 +1,52 @@
...
 	public Map<String, String> getHeader(@RequestHeader Map<String, String> headers) {
 		return headers;
 	}
+
+	@PostMapping("/schedule")
+	public ResponseEntity<?> schedule() {
+
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java:34]: use proper formatting. remove the unnecessary white line. I would recommend reading about the eclipse java formatting rule once.

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 5]
```diff
@@ -1,16 +1,52 @@
...
+	@Autowired
+	ScheduleService service;
+
+	@GetMapping("/header")
 	public Map<String, String> getHeader(@RequestHeader Map<String, String> headers) {
 		return headers;
 	}
+
+	@PostMapping("/schedule")
+	public ResponseEntity<?> schedule() {
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]:
```suggestion
	public ResponseEntity<String> schedule() {
```
don't use wildcard here

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 6]
```diff
@@ -1,16 +1,52 @@
...
+		ExecutorService executor = Executors.newSingleThreadExecutor();
+
+		// process request in other thread
+		executor.submit(() -> service.processRequest(5));
+
+		return new ResponseEntity<>("SUCCESS", HttpStatus.OK);
+	}
+
+	@ExceptionHandler(Exception.class)
+	public final ResponseEntity<?> handleAllExceptions(Exception ex) {
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]: better create a separate exception handler class.

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 7]
```diff
@@ -1,16 +1,52 @@
...
+		log.error("@ExceptionHandler: ", ex);
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]: handle logs as we discussed, use parameter formatting.

File: src/main/java/com/staffmanagement/controller/UtilityController.java [hunk 8]
```diff
@@ -1,16 +1,52 @@
...
+		return new ResponseEntity<>(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/controller/UtilityController.java]:
```suggestion
		return new ResponseEntity<String>(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
```

File: src/main/java/com/staffmanagement/service/ScheduleService.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.context.annotation.ScopedProxyMode;
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope(value = "prototype", proxyMode = ScopedProxyMode.TARGET_CLASS)
+public class ScheduleService {
+	Logger log = LoggerFactory.getLogger(ScheduleService.class);
```
Comment (rvashishth) [inline src/main/java/com/staffmanagement/service/ScheduleService.java]: these should be private

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -2,4 +2,9 @@ server.port=9090
 
 server.servlet.context-path=/staff
 
+logging.file.max-size:10MB
+logging.file.max-history=2
+logging.file=E:\myworkspace\staff_management\log.txt
```
Comment (rvashishth) [inline src/main/resources/application.properties]: it would be more appropriate to move this file property to environment-specific properties.

**Review by rvashishth — CHANGES_REQUESTED** [review]
A suggestion - always use sonarlint.  Later we should be setting up sonarcloud using github actions

File: src/main/java/com/staffmanagement/controller/UtilityController.java (hunk 4 above)
Comment (Poonam00) [reply src/main/java/com/staffmanagement/controller/UtilityController.java:34]: Whitespace and new lines are crucial for clean readable code.

**Review by rvashishth — APPROVED** [review]

**Pull request merged**